services:
  grafana:
    image: grafana/grafana:latest
    ports:
      - "3001:3000" #map Grafana container's 3000 port to host's 3001
    volumes:
      - ./grafana/provisioning:/etc/grafana/provisioning
      - grafana-storage:/var/lib/grafana
    environment:
      - GF_SECURITY_ADMIN_USER=admin
      - GF_SECURITY_ADMIN_PASSWORD=adminpassword
    depends_on:
      - postgres
    networks:
      - app-network

  postgres:
    image: postgres:latest
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_DB: mydb
      TZ: Asia/Kolkata  # Change DBeaver timezone settings
    volumes:
      - ./postgres:/docker-entrypoint-initdb.d
      - postgres-data:/var/lib/postgresql/data1     #if there are initialization errors after multiple runs --> docker rm -f $(docker ps -a -q),  docker volume rm customapp_postgres-data
    ports:
      - "5434:5432"
    healthcheck:
      test: ["CMD", "pg_isready", "-U", "postgres"]
      interval: 10s
      timeout: 5s
      retries: 5
    networks:
      - app-network

  python:     # if there are older builds, then the new packages in requirements.txt won't be installed. Run "docker compose build --no-cache"
    build:
      context: ./Custom-App
      dockerfile: Dockerfile
    ports:
      - "8000:8000"  # FastAPI
    volumes:
      - ./Custom-App/app:/app
    depends_on:
      - postgres
    networks:
      - app-network
    command: >
      bash -c "uvicorn KiteAccessToken:app --host 0.0.0.0 --port 8000 &
      python KiteWS.py &  
      wait"   

  redis:
    image: redis:7-alpine
    restart: unless-stopped
    ports:
      - "127.0.0.1:6379:6379"
    networks:
      - app-network
    volumes:
      - redis-data:/data

volumes:
  grafana-storage:
  postgres-data:
  redis-data:

networks:
  app-network:
    driver: bridge
